Add DO cluster creation and teardown to bootstrap instructions (#51)
vipyrsec/infrastructure: docs/bootstrap.md

# Bootstrapping Dragonfly on Digital Ocean

## Prerequisites

You will need:

* [`doctl`](https://docs.digitalocean.com/reference/doctl/how-to/install/) configured with a...
* [Digital Ocean Personal Access Token](https://docs.digitalocean.com/reference/api/create-personal-access-token/) with at least the following scopes:
  * kubernetes (2): create, delete
  * If replacing an existing cluster:

>     * load\_balancer (1): delete

Reviewer: `question`: We'll need this as well, right?
```md
kubernetes: access_cluster
```

Author: That scope only controls access to these two endpoints, which I'm pretty sure we don't (technically) need:
* https://docs.digitalocean.com/reference/api/api-reference/#operation/kubernetes_get_kubeconfig
* https://docs.digitalocean.com/reference/api/api-reference/#operation/kubernetes_get_credentials

Reviewer: Okay, we can get those from the UI, I think.

> * PWD set to the root of this repo

Reviewer: `suggestion(non-blocking)`:

```suggestion
* `PWD` set to the root of this repo
```

* [`kubectl`](https://kubernetes.io/docs/tasks/tools/#kubectl)
* [`helm`](https://helm.sh/docs/intro/install/)

## Create the cluster in Digital Ocean

> ```

Reviewer: `suggestion(non-blocking)`:

```suggestion
```bash
```

doctl k8s cluster create \
  <cluster-name> \
  --size s-2vcpu-2gb \
  --count 2 \
  --region nyc3
```

> Adjust `size` and `count` as needed. See [`doctl kubernetes cluster create
> docs`](https://docs.digitalocean.com/reference/doctl/reference/kubernetes/cluster/create/)
> for more options.

Reviewer: `suggestion(blocking)`: We could use another `region`.

```suggestion
Adjust `size`, `count`, and `region` as needed. See [`doctl kubernetes cluster create` docs](https://docs.digitalocean.com/reference/doctl/reference/kubernetes/cluster/create/)for more options.
```

> ## Apply `cert-manager` CRDs

Reviewer: `suggestion(blocking)`: Add a step to make sure the kube context is set to the new cluster.

Author: Creating the cluster using `doctl` does this implicitly. Could add a note to reassure readers that this is happening, though the `doctl` output does say this I think.

> ```

Reviewer: `suggestion(non-blocking)`:

```suggestion
```bash
```

> kubectl apply -f https://github.com/cert-manager/cert-manager/releases/download/v1.12.0/cert-manager.crds.yaml

Reviewer: `suggestion(blocking)`: We actually commit these, so we can use those.

```suggestion
kubectl apply -f ./kubernetes/crds/cert-manager.crds.yaml
```

```

## Install the Helm Chart for `ingress-nginx` and `cert-manager`

```bash
helm upgrade vipyrsec ./kubernetes/chart --install
```

## Apply manifests and secrets

> Secrets are not included in this repo; you will need to create your own.

Reviewer: `suggestion(blocking)`: Moving the manifests to the Helm chart will solve this, but for now, it is what it is.

```suggestion
You might need to run the below command twice, since the namespaces might not get created before the other manifests depending on them.
Secrets are not included in this repo, so you will need to create those separately.
```

> ```

Reviewer: `suggestion(non-blocking)`:

```suggestion
```bash
```

kubectl apply -f ./kubernetes/manifests -R
```

## If access over the internet is required, create a DNS A Record to use the new load balancer

## If replacing an existing cluster, destroy old resources

> ```

Reviewer: `suggestion(non-blocking)`:

```suggestion
```bash
```

doctl k8s cluster delete <name>

> doctl compute load-balancer delete <id>

Reviewer: `suggestion(non-blocking)`: Getting the ID is an unnecessary step.

```suggestion
doctl compute load-balancer delete <name>
```

Author: Hm, this docs page says you need `<id>`, https://docs.digitalocean.com/reference/doctl/reference/compute/load-balancer/delete/.

Reviewer: Huh, weird. Where do you get the ID from? The UI?

  * kubernetes: create, delete
  * If replacing an existing cluster:
    * load\_balancer: read, delete
* `PWD` set to the root of this repo
* [`kubectl`](https://kubernetes.io/docs/tasks/tools/#kubectl)
* [`helm`](https://helm.sh/docs/intro/install/)

## Create the cluster in Digital Ocean
```bash
doctl k8s cluster create \
  <cluster-name> \
  --size s-2vcpu-2gb \
  --count 2 \
  --region nyc3
```

Adjust `size`, `count`, and `region` as needed. See [`doctl kubernetes cluster
create
docs`](https://docs.digitalocean.com/reference/doctl/reference/kubernetes/cluster/create/)
for more options. The cluster may take a few minutes to provision.

**Note**: `doctl k8s cluster create` sets the `kubectl` context to the newly
created cluster.

## Apply `cert-manager` CRDs

```bash
kubectl apply -f ./kubernetes/crds/cert-manager.crds.yaml
```

## Install the Helm Chart for `ingress-nginx` and `cert-manager`

```bash
helm upgrade vipyrsec ./kubernetes/chart --install
```

## Apply manifests and secrets

Secrets are not included in this repo; you will need to create your own.

```bash

> kubectl apply -f kubernetes\manifests\cert-manager\namespace.yaml
> kubectl apply -f kubernetes\manifests\discord\namespace.yaml
> kubectl apply -f kubernetes\manifests\dragonfly\namespace.yaml

Reviewer: suggestion(non-blocking):

```suggestion
kubectl apply -f ./kubernetes/manifests/cert-manager/namespace.yaml
kubectl apply -f ./kubernetes/manifests/discord/namespace.yaml
kubectl apply -f ./kubernetes/manifests/dragonfly/namespace.yaml
```
